src/imitation/rewards/reward_nets.py

```diff
@@ -409,6 +407,218 @@ def forward(self, state, action, next_state, done):
+        self.use_done = use_done
+        self.input_size = 0
+        self.obs_is_image = preprocessing.is_image_space(observation_space)
+        self.act_is_box_with_channels = (
+            isinstance(action_space, spaces.Box) and len(action_space.shape) == 3
+        )
+        self.hwc_format = hwc_format
+
+        if not self.can_handle_space(observation_space):
+            raise ValueError(
+                "CnnRewardNet doesn't know how to deal with this observation space.",
+            )
+        if not self.can_handle_space(action_space):
+            raise ValueError(
+                "CnnRewardNet doesn't know how to deal with this action space.",
+            )
+
+        if self.use_state:
+            self.input_size += self.get_num_channels(observation_space)
+
+        if self.use_action:
+            self.input_size += self.get_num_channels(action_space)
+
+        if self.use_next_state:
+            self.input_size += self.get_num_channels(observation_space)
+
+        if self.use_done:
+            self.input_size += 1
+
+        full_build_cnn_kwargs = {
+            "hid_channels": (32, 32),
+            **kwargs,
+            # we do not want the values below to be overridden
+            "in_channels": self.input_size,
+            "out_size": 1,
+            "squeeze_output": True,
+        }
+
+        self.cnn = networks.build_cnn(**full_build_cnn_kwargs)
+
+    def can_handle_space(self, space: gym.Space) -> bool:
+        """Tells us if this CNN can handle the given gym space."""
+        space_is_ok = isinstance(
+            space,
+            (spaces.Box, spaces.Discrete, spaces.MultiDiscrete, spaces.MultiBinary),
+        )
+        if isinstance(space, spaces.MultiDiscrete):
+            shape_is_ok = len(space.nvec.shape) == 1
+        elif isinstance(space, spaces.MultiBinary):
+            shape_is_ok = isinstance(space.n, int)
+        else:
+            shape_is_ok = True
+        return space_is_ok and shape_is_ok
+
+    def get_num_channels(
+        self,
+        space: Union[
+            spaces.Box,
+            spaces.Discrete,
+            spaces.MultiDiscrete,
+            spaces.MultiBinary,
+        ],
+    ) -> int:
+        """Gets number of channels for the representation of a gym space."""
+        if isinstance(space, spaces.Box):
+            my_dim = space.shape[-1] if self.hwc_format else space.shape[0]
+        elif isinstance(space, spaces.Discrete):
+            my_dim = space.n
+        elif isinstance(space, spaces.MultiDiscrete):
+            my_dim = sum(space.nvec)
+        elif isinstance(space, spaces.MultiBinary):
+            my_dim = space.n if isinstance(space.n, int) else sum(space.n)
+        else:
+            raise TypeError(
+                "get_num_channels can't recognize the input space. "
+                + "Please provide one of Discrete, MultiDiscrete, MultiBinary, Box.",
+            )
+
+        return my_dim
+
+    def transpose(
```

Comment:
I was worried that it might happen somewhere, but if you think it rarely does then I can make these changes.

Reply:
I don't think it does in any of our algorithms. If I understand https://github.com/HumanCompatibleAI/imitation/blob/master/src/imitation/rewards/reward_nets.py#L407 correctly, then our existing `BasicRewardNet` would break if there's no batch dimension.

Reply:
Indeed, if I `assert False` in the branch where there is no batch dimension, tests still pass and the notebook still runs. Will rewrite to explicitly assume that.